src/Execution/Arguments/ArgPartitioner.php

```diff
@@ -0,0 +1,204 @@
+use Illuminate\Database\Eloquent\Relations\MorphMany;
+use Illuminate\Database\Eloquent\Relations\MorphToMany;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+
+class ArgPartitioner
+{
+    /**
+     * Partition the arguments into nested and regular.
+     *
+     * @param  \Nuwave\Lighthouse\Execution\Arguments\ArgumentSet  $argumentSet
+     * @param  mixed  $root
+     * @return \Nuwave\Lighthouse\Execution\Arguments\ArgumentSet[]
+     */
+    public static function nestedArgumentResolvers(ArgumentSet $argumentSet, $root): array
+    {
+        $model = $root instanceof Model
+            ? new \ReflectionClass($root)
+            : null;
+
+        foreach ($argumentSet->arguments as $name => $argument) {
+            static::attachNestedArgumentResolver($name, $argument, $model);
+        }
+
+        return static::partition(
+            $argumentSet,
+            static function (string $name, Argument $argument): bool {
+                return isset($argument->resolver);
+            }
+        );
+    }
+
+    /**
+     * Extract all the arguments that correspond to a relation of a certain type on the model.
+     *
+     * For example, if the args input looks like this:
+     *
+     * [
+     *  'name' => 'Ralf',
+     *  'comments' =>
+     *    ['foo' => 'Bar'],
+     * ]
+     *
+     * and the model has a method "comments" that returns a HasMany relationship,
+     * the result will be:
+     * [
+     *   [
+     *    'comments' =>
+     *      ['foo' => 'Bar'],
+     *   ],
+     *   [
+     *    'name' => 'Ralf',
+     *   ]
+     * ]
+     *
+     * @param  \Nuwave\Lighthouse\Execution\Arguments\ArgumentSet  $argumentSet
+     * @param  \Illuminate\Database\Eloquent\Model  $model
+     * @param  string  $relationClass
+     * @return \Nuwave\Lighthouse\Execution\Arguments\ArgumentSet[]
+     */
+    public static function relationMethods(
+        ArgumentSet $argumentSet,
+        Model $model,
+        string $relationClass
+    ): array {
+        $modelReflection = new ReflectionClass($model);
+
+        return static::partition(
+            $argumentSet,
+            static function (string $name) use ($modelReflection, $relationClass): bool {
+                return static::methodReturnsRelation($modelReflection, $name, $relationClass);
+            }
+        );
+    }
+
+    /**
+     * Attach a nested argument resolver to an argument.
+     *
+     * @param  string  $name
+     * @param  \Nuwave\Lighthouse\Execution\Arguments\Argument  $argument
+     * @param  \ReflectionClass|null  $model
+     * @return void
+     */
+    protected static function attachNestedArgumentResolver(string $name, Argument &$argument, ?ReflectionClass $model): void
+    {
+        $resolverDirective = $argument->directives->first(function (Directive $directive): bool {
+            return $directive instanceof ArgumentResolver;
+        });
+
+        if ($resolverDirective) {
```

Comment:
What do you think about to add a ArgumentMiddleware that allows you to execute previous actions before the resolver?

I think this should be easy to implement just checking it here and wrapping the `$argument->resolver`. Actually you are already using the middleware pattern here with thinks like `new ResolveNested(new UpsertModel(new SaveModel($relation)))`. It could/should be another PR, ofc.

Reply:
Hey @joskfg i am finally coming around to finishing up this PR. Your idea is a good one, thanks. We can do it in another PR.